[NCCL][Blocking Wait] Log set exceptions when checking for exceptions in
blocking wait

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54742 test_c10d: use with_nccl_blocking_wait decorator
* #54741 test_c10d: Run tests with nccl_async_error_handling
* #54740 Provide a decorator to set/unset nccl blocking wait for tests
* **#54558 [NCCL][Blocking Wait] Log set exceptions when checking for exceptions in
blocking wait**
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier
blocking wait**
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier
blocking wait**
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier
blocking wait**
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier
blocking wait**
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier
blocking wait**
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier
blocking wait**
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier
blocking wait**
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier

blocking wait

In blocking wait's polling synchronization loop, we frequently call checkAndSetException() as part of isCompleted() to check the status of nccl operations. It would be useful to log here in case we encounter any exceptions (which are later thrown by `checkAndThrowException`).

Also slightly refactors code previously added to make use of a helper function to get the error message given an `std::exception_ptr`.

Differential Revision: [D27136202](https://our.internmc.facebook.com/intern/diff/D27136202/)

rohan-varma committed:
[NCCL][Blocking Wait] Log set exceptions when checking for exceptions in
blocking wait

In blocking wait's polling synchronization loop, we frequently call checkAndSetException() as part of isCompleted() to check the status of nccl operations. It would be useful to log here in case we encounter any exceptions (which are later thrown by `checkAndThrowException`).

Also slightly refactors code previously added to make use of a helper function to get the error message given an `std::exception_ptr`.

Differential Revision: [D27136202](https://our.internmc.facebook.com/intern/diff/D27136202/)

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
I think you can rewrite this as follows:

```
    try {
      std::rethrow_exception(eptr);
    } catch (const std::exception& e) {
      return e.what();
    } catch (...) {
      return "Unknown Exception Type";
    }
```

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
Do we need to LOG here again for the same rank? We would already have a LOG when we set the exception for the first time below right?

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
nit: can just call `getExceptionMsgFromExceptionPtr(exception_);` inline here?

rohan-varma commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
That's a good point, we've already logged this one and there's no reason to log it again.

rohan-varma committed:
Update on "[NCCL][Blocking Wait] Log set exceptions when checking for exceptions in
blocking wait"

blocking wait**
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier

blocking wait

In blocking wait's polling synchronization loop, we frequently call checkAndSetException() as part of isCompleted() to check the status of nccl operations. It would be useful to log here in case we encounter any exceptions (which are later thrown by `checkAndThrowException`).

Also slightly refactors code previously added to make use of a helper function to get the error message given an `std::exception_ptr`.

Differential Revision: [D27136202](https://our.internmc.facebook.com/intern/diff/D27136202/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[NCCL][Blocking Wait] Log set exceptions when checking for exceptions in
blocking wait"

blocking wait**
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier
blocking wait**
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier

blocking wait

In blocking wait's polling synchronization loop, we frequently call checkAndSetException() as part of isCompleted() to check the status of nccl operations. It would be useful to log here in case we encounter any exceptions (which are later thrown by `checkAndThrowException`).

Also slightly refactors code previously added to make use of a helper function to get the error message given an `std::exception_ptr`.

Differential Revision: [D27136202](https://our.internmc.facebook.com/intern/diff/D27136202/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[NCCL][Blocking Wait] Log set exceptions when checking for exceptions in
blocking wait"

blocking wait**
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier
blocking wait**
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier
blocking wait**
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier
blocking wait**
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier

blocking wait

In blocking wait's polling synchronization loop, we frequently call checkAndSetException() as part of isCompleted() to check the status of nccl operations. It would be useful to log here in case we encounter any exceptions (which are later thrown by `checkAndThrowException`).

Also slightly refactors code previously added to make use of a helper function to get the error message given an `std::exception_ptr`.

Differential Revision: [D27136202](https://our.internmc.facebook.com/intern/diff/D27136202/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[NCCL][Blocking Wait] Log set exceptions when checking for exceptions in
blocking wait"

blocking wait**
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier
blocking wait**
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier
blocking wait**
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier
blocking wait**
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier
blocking wait**
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier

blocking wait

In blocking wait's polling synchronization loop, we frequently call checkAndSetException() as part of isCompleted() to check the status of nccl operations. It would be useful to log here in case we encounter any exceptions (which are later thrown by `checkAndThrowException`).

Also slightly refactors code previously added to make use of a helper function to get the error message given an `std::exception_ptr`.

Differential Revision: [D27136202](https://our.internmc.facebook.com/intern/diff/D27136202/)

[ghstack-poisoned]
